### PR TITLE
gh action reworked to publish image before mvn verify job

### PR DESCRIPTION
**Pull request by rigazilla**
Build / Test workflow has been split in two different, so the publish image can be triggered ad the end of the build workflow.

### CODE REVIEW

File: .github/workflows/on_build_do_publish_image.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      sourceHeadSha: ${{ steps.workflow-run-info.outputs.sourceHeadSha }}
+      mergeCommitSha: ${{ steps.workflow-run-info.outputs.mergeCommitSha }}
+      targetCommitSha: ${{ steps.workflow-run-info.outputs.targetCommitSha }}
+      pullRequestNumber: ${{ steps.workflow-run-info.outputs.pullRequestNumber }}
+      pullRequestLabels: ${{ steps.workflow-run-info.outputs.pullRequestLabels }}
+      targetBranch: ${{ steps.workflow-run-info.outputs.targetBranch }}
+      sourceEvent: ${{ steps.workflow-run-info.outputs.sourceEvent }}
+    steps:
+      - name: "Get information about the current run"
+        uses: potiuk/get-workflow-origin@v1_1
```
Comment (ryanemerson) [inline .github/workflows/on_build_do_publish_image.yml:30]: This action has been archived since 2023, I don't think it's a good idea to rely on a dead project.

File: .github/workflows/on_build_do_test.yml [hunk 2]
```diff
@@ -32,6 +42,8 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+        with:
+         ref: ${{ needs.get-info.outputs.sourceHeadSha }}
```
[the remark this answers is not in the document]
Comment (rigazilla) [reply .github/workflows/on_build_do_test.yml:46]: uh good spot!, test should run with restriction (just like on pull_request)... I'll fix this

File: .github/workflows/on_build_do_publish_image.yml (hunk 1 above)
Comment (rigazilla) [reply .github/workflows/on_build_do_publish_image.yml:30]: yeah, this can be converted in API calls
I don't know why they decided to archive it, I find it pretty useful

File: .github/workflows/on_build_do_test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rigazilla) [reply .github/workflows/on_build_do_test.yml:46]: I'm going to wrap the mvn verify step in an composite action so secrets are unaccessible and set permission at the same level as pull_request

[the remark this answers is not in the document]
Comment (rigazilla) [reply .github/workflows/on_build_do_test.yml]: now using src.zip artifact to run test

File: .github/workflows/on_build_do_publish_image.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      uses: infinispan/infinispan-images/.github/workflows/publish_image.yml@main
+      secrets:
+         token: ${{ secrets.GITHUB_TOKEN }}
+         quayUser: ${{ secrets.QUAY_USERNAME }}
+         quayPass: ${{ secrets.QUAY_TOKEN }}
+      with:
+         repository: ${{ github.repository }}
+         runId: ${{ github.event.workflow_run.id }}
+         serverArtifact: infinispan-dist
+         tags:  ${{ needs.get-server-version.outputs.imageName }}:gh-${{ needs.get-info.outputs.sourceHeadBranch }}
```
Comment (ryanemerson) [inline .github/workflows/on_build_do_publish_image.yml:66]: AFAICT the only difference between these two jobs is the `tags` field. We can calculate the tags as part of the `get-server-version` job, or a dedicated job and then use the output so that we only need a single job fro both `push` and `pull-request` events.

File: .github/workflows/on_build_do_test.yml [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            -Dmaven.test.failure.ignore=true -Dansi.strip=true -Pnative -pl commons/all
+
+      - name: Archive commit sha PR
+        if: >
+          (success() || failure()) &&
+          needs.get-info.outputs.sourceEvent == 'pull_request'
+        run: |
+           echo -n ${{ needs.get-info.outputs.sourceHeadSha }} > github-sha.txt
+
+      - name: Archive commit sha push
```
Comment (ryanemerson) [inline .github/workflows/on_build_do_test.yml:99]: I don't think we need to create these sha files any more?

**Review by ryanemerson — COMMENTED** [review]
Thanks for addressing the security concern @rigazilla.

Looking through the .yml, my first thought is that this is getting very verbose and hard to follow with the various files that are passed etc and the use of `workflow_run` to protect secrets. I fear that this is only going to get worse as we embark on adopting GH issues and the various automations that will be required.

Is it time to reconsider a bot to handle some of this?

In this scenario, we could do the following:

1. Workflow builds Infinispan, uploads .zip and adds a comment to the PR if an image is required (it might be possible to get an event when an artifact is uploaded, I haven't checked)

2. Bot detects the "Image required" label on PR upate event and then executes a build Image workflow

This way there's no need to upload sources etc to workaround secret access.

**Comment by rigazilla** [comment]
> Thanks for addressing the security concern @rigazilla.
> 
> Looking through the .yml, my first thought is that this is getting very verbose and hard to follow with the various files that are passed etc and the use of `workflow_run` to protect secrets. I fear that this is only going to get worse as we embark on adopting GH issues and the various automations that will be required.
> 
> Is it time to reconsider a bot to handle some of this?
>
mmm don't know, the separation between unsafe stuff (running PR code) and the rest seems done to me, and I consider it the hard part. 

tbh, if this works, I would spend more effort in trying to make the whole process more readable and adding new features instead of redoing same things in another way adding more pieces. just my opinion
 
> In this scenario, we could do the following:
> 
>     1. Workflow builds Infinispan, uploads .zip and adds a comment to the PR if an image is required (it might be possible to get an event when an artifact is uploaded, I haven't checked)

unfortunately seems that there aren't events triggered by artifacts, this made all the thing more complicated

> 
>     2. Bot detects the "Image required" label on PR upate event and then executes a build Image workflow
> 
> 
> This way there's no need to upload sources etc to workaround secret access.

File: .github/workflows/on_build_do_test.yml (hunk 4 above)
Comment (rigazilla) [reply .github/workflows/on_build_do_test.yml:99]: is needed to attach the number-of-test-ran check to the correct commit:
https://github.com/rigazilla/infinispan/blob/4f736c33aee2da21356b3772a7ac2fb933533bdf/.github/workflows/on_test_publish_surefire_report.yml#L86

File: .github/workflows/on_build_do_publish_image.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
+name: On Build Do Publish Image
+on:
+   workflow_run:
+      workflows: [On PR Open or Push Do Build]
+      types: [completed]
```
Comment (ryanemerson) [inline .github/workflows/on_build_do_publish_image.yml:5]: Is the image created for every PR?

The current Jenkins images are only created when a PR is annotated with "Image Required". We don't currently have any automated mechanism in place to cleanup old quay.io/infinispan-test/server images, but I think this will be required if we create an image per PR, although I don't know what quay.io usage limits are.

Comment (rigazilla) [reply .github/workflows/on_build_do_publish_image.yml:5]: It's ok to reintroduce the label. I was thinking to do it in the next step, but better to add it from the beginning... no need to fill quay with dependabot images

File: .github/workflows/on_propen_push_do_build.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Build Infinispan
+        uses: ./.github/actions/build-infinispan
+
+      - run: echo Build Completed ${{ github.run_id }} > server-version.txt
```
Comment (ryanemerson) [inline .github/workflows/on_propen_push_do_build.yml:25]: Do we really need to upload this file. Can't we extract the version of the server based upon the pattern in the server/src path?

Comment (rigazilla) [reply .github/workflows/on_propen_push_do_build.yml:25]: > Do we really need to upload this file. Can't we extract the version of the server based upon the pattern in the server/src path?
> 
> We could even go further and have a `./github/actions/download-infinispan` that extracts the server and src zips, then sets the server version as an output that can be consumed by subsequent steps. This should reduce the verbosity of some of our workflows and centralise duplicated logic.

do you mean from the file name in the artifact? If it's a reliable info sure we can

Comment (ryanemerson) [reply .github/workflows/on_propen_push_do_build.yml:25]: Yep. That version number is dictated by the project.version, so it's effectively the same as what the action currently does.

File: commons/all/src/test/java/org/infinispan/commons/util/HashMapTest.java [hunk 7]
```diff
@@ -21,6 +21,21 @@ public void testCreateHopscotchWithRemovals() {
...
+
+   public static int i=0;
+   @Test
+   public void testFlaky() {
+      if (i==0) {
+         i=1;
+         assertEquals(1,2);
+      }
+      assertEquals(1,1);
+   }
```
Comment (fax4ever) [inline commons/all/src/test/java/org/infinispan/commons/util/HashMapTest.java:37]: Maybe we should move those tests in a different test class. If I understand correctly, the idea here is to simulate failing an flaky tests.

**Review by fax4ever — COMMENTED** [review]
Great job. One question:
Is there a way to verify that we can upload the logs of the execution.
For instance `infinispan-infinispan-client-hotrod.log`?

File: commons/all/src/test/java/org/infinispan/commons/util/HashMapTest.java (hunk 7 above)
Comment (rigazilla) [reply commons/all/src/test/java/org/infinispan/commons/util/HashMapTest.java]: yeah I need them just for dev, I'll cleanup in the final version

**Comment by rigazilla** [comment]
I excluded RocksDBReaderTest surefire report since sometimes it's broken xml

File: .github/workflows/on_build_do_publish_image.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+          name: Get Labels Action
+          uses: snnaplab/get-labels-action@v1.0.1
+          with:
+            number: ${{ steps.workflow-run-info.outputs.pull-request-number }}
+
+        - if: ${{ github.event.workflow_run.event == 'pull_request' && contains(fromJSON(env.LABELS), 'Image Required') }}
+          id: publish-image-info
+          run: echo "publish-image=true" >> $GITHUB_OUTPUT
+
+  get-image-name:
```
Comment (ryanemerson) [inline .github/workflows/on_build_do_publish_image.yml:44]: Can't this just be a step in the `get-info` job?

**Review by ryanemerson — COMMENTED** [review]
One minor comment, otherwise LGTM

File: .github/workflows/on_build_do_publish_image.yml (hunk 8 above)
Comment (rigazilla) [reply .github/workflows/on_build_do_publish_image.yml]: actually yes

**Comment by ryanemerson** [comment]
Thanks @rigazilla
